lib/db_utils.go

```diff
@@ -8974,3 +9139,89 @@ func DBDeletePostAssociationWithTxn(txn *badger.Txn, snap *Snapshot, association
 	}
 	return nil
 }
+
+func EnumerateKeysForPrefixWithLimitOffsetOrder(
+	db *badger.DB,
+	prefix []byte,
+	limit int,
+	lastSeenKey []byte,
+	sortDescending bool,
+	deletedUtxoKeys *Set[string],
+) ([][]byte, [][]byte, error) {
+	keysFound := [][]byte{}
+	valsFound := [][]byte{}
+
+	dbErr := db.View(func(txn *badger.Txn) error {
+		var err error
+		keysFound, valsFound, err = _enumerateKeysForPrefixWithLimitOffsetOrderWithTxn(
+			txn, prefix, limit, lastSeenKey, sortDescending, deletedUtxoKeys,
+		)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if dbErr != nil {
+		return nil, nil, errors.Wrapf(
+			dbErr,
+			"EnumerateKeysForPrefixWithLimitOffsetOrder: problem fetching keys and values from db: ",
+		)
+	}
+
+	return keysFound, valsFound, nil
+}
+
+func _enumerateKeysForPrefixWithLimitOffsetOrderWithTxn(
+	txn *badger.Txn,
+	prefix []byte,
+	limit int,
+	lastSeenKey []byte,
+	sortDescending bool,
+	deletedUtxoKeys *Set[string],
+) ([][]byte, [][]byte, error) {
+	keysFound := [][]byte{}
+	valsFound := [][]byte{}
+
+	// If provided, start at the last seen key.
+	startingKey := prefix
+	if lastSeenKey != nil {
+		startingKey = lastSeenKey
+	}
+
+	opts := badger.DefaultIteratorOptions
+	// Search keys in reverse order if sort DESC.
+	if sortDescending {
+		opts.Reverse = true
+		startingKey = append(startingKey, 0xff)
+	}
+	nodeIterator := txn.NewIterator(opts)
+	defer nodeIterator.Close()
+
+	for nodeIterator.Seek(startingKey); nodeIterator.ValidForPrefix(prefix); nodeIterator.Next() {
+		// Break if at or beyond limit.
+		if limit > 0 && len(keysFound) >= limit {
+			break
+		}
+		// Copy key.
+		key := nodeIterator.Item().Key()
+		keyCopy := make([]byte, len(key))
+		copy(keyCopy[:], key[:])
+		// Skip if key is the last seen key.
+		if bytes.Equal(keyCopy, lastSeenKey) {
+			continue
```

Comment:
should we track whether or not we've seen the `lastSeenKey` so we can avoid using `bytes.Equal` on future keys?